File: Tutorials/BiometricStandards/Java/fmcrecord-to-ntemplate/src/main/java/com/neurotec/tutorials/FMCRecordToNTemplate.java
package com.neurotec.tutorials;

import java.io.IOException;

import com.neurotec.biometrics.NFRecord;
import com.neurotec.biometrics.NFTemplate;
import com.neurotec.biometrics.NTemplate;
import com.neurotec.biometrics.standards.BDIFStandard;
import com.neurotec.biometrics.standards.FMCRMinutiaFormat;
import com.neurotec.biometrics.standards.FMCRecord;
import com.neurotec.io.NBuffer;
import com.neurotec.io.NFile;
import com.neurotec.licensing.NLicense;
import com.neurotec.licensing.NLicenseManager;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;
import com.neurotec.util.NVersion;

public final class FMCRecordToNTemplate {

	private static final String DESCRIPTION = "Converting FMCRecord to NTemplate";
	private static final String NAME = "fmcrecord-to-ntemplate";
	private static final String VERSION = "13.1.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t %s [FMCRecord] [NTemplate] [Version] [MinutiaeFormat] [ReadBerTlv]%n", NAME);
		System.out.println();
		System.out.println("\t[FMCRecord] - filename of FMCRecord.");
		System.out.println("\t[NTemplate] - filename of NTemplate to be created.");
		System.out.println("\t[Version] - FMCRecord version.");
		System.out.println("\t\tISO2 - ISO/IEC 19794-2:2005.");
		System.out.println("\t\tISO3 - ISO/IEC 19794-2:2011.");
		System.out.println("\t[MinutiaeFormat] - card format of minutiae data.");
		System.out.println("\t\tC - compact size.");
		System.out.println("\t\tN - normal size (ISO2 only).");
		System.out.println("\tReadBerTlvDo - FMCRecord read from buffer option");
		System.out.println("\t\t1 - Biometric Data Template (BDT) or Cardholder");
		System.out.println("\t\t    Biometric Data (CBD) BER-TLV DO (Data Object)");
		System.out.println("\t\t0 - minutiae data buffer only.\n");
		System.out.println();
		System.out.println("example:");
		System.out.format("\t%s fmcRecord.FMCRecord template.NTemplate ISO3 C 1%n", NAME);
	}

	public static void main(String[] args) {
		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 5) {
			usage();
			System.exit(1);
		}

		//=========================================================================
		// CHOOSE LICENCES !!!
		//=========================================================================
		// ONE of the below listed licenses is required for unlocking this sample's functionality. Choose a license that you currently have on your device.
		// If you are using a TRIAL version - choose any of them.

		final String license = "FingerClient";
		//final String license = "FingerFastExtractor";

		//=========================================================================

		//=========================================================================
		// TRIAL MODE
		//=========================================================================
		// Below code line determines whether TRIAL is enabled or not. To use purchased licenses, don't use below code line.
		// GetTrialModeFlag() method takes value from "Bin/Licenses/TrialFlag.txt" file. So to easily change mode for all our examples, modify that file.
		// Also you can just set TRUE to "TrialMode" property in code.
		//=========================================================================

		try {
			boolean trialMode = Utils.getTrialModeFlag();
			NLicenseManager.setTrialMode(trialMode);
			System.out.println("\tTrial mode: " + trialMode);
		} catch (IOException e) {
			e.printStackTrace();
		}

		FMCRecord fmcRecord = null;
		NTemplate nTemplate = null;
		NBuffer storedFmcRecord = null;
		NBuffer packedNTemplate = null;

		try {
			// Obtain license
			if (!NLicense.obtain("/local", 5000, license)) {
				System.err.format("Could not obtain license: %s%n", license);
				System.exit(-1);
			}

			BDIFStandard standard = BDIFStandard.ISO;
			int flags = 0; // BDIFTypes.FLAG_NON_STRICT_READ - removes non-unique minutiae for FMCRecord.VERSION_ISO_30 only

			NVersion standardVersion;
			if (args[2].equals("ISO2")) {
				standardVersion = FMCRecord.VERSION_ISO_20;
			} else if (args[2].equals("ISO3")) {
				standardVersion = FMCRecord.VERSION_ISO_30;
			} else {
				throw new Exception("Wrong standard");
			}

			FMCRMinutiaFormat minutiaFormat;
			if (args[3].equals("C")) {
				minutiaFormat = FMCRMinutiaFormat.COMPACT_SIZE;
			} else if (args[3].equals("N")) {
				minutiaFormat = FMCRMinutiaFormat.NORMAL_SIZE;
			} else {
				throw new Exception("Wrong minutia format");
			}

			// Check if full FMCRecord BER-TLV Data Object (DO) (e.g., Biometric Data Template (BDT) or Cardholder Biometric Data (CBD)) or minutiae buffer only is available
			boolean isBerTlv = args[4].equals("1");

			storedFmcRecord = NFile.readAllBytes(args[0]);
			if (isBerTlv) {
				// Create FMCRecord object from FMCRecord BER-TLV Data Object (DO) stored in memory (e.g., BDT or CBD BER-TLV DO)
				fmcRecord = new FMCRecord(storedFmcRecord, standard, standardVersion, minutiaFormat, flags);
			} else { // if (!isBerTlv)
				// Create FMCrecord and read minutiae data from buffer to FMCRecord
				fmcRecord = new FMCRecord(standard, standardVersion, minutiaFormat);
				fmcRecord.setMinutiaeBuffer(storedFmcRecord);
			}

			// Explicitly check minutiae uniqueness for FMCRecord v2.0 as well (since mandatory implicit check since v3.0 only)
			if (fmcRecord.getVersion().equals(FMCRecord.VERSION_ISO_20) && !fmcRecord.isValidateMinutiaeUniqueness()) {
				System.out.println("!!!WARNING!!!:");
				System.out.println("Not all minutiae in FMCRecord are unique!");
				System.out.println(
						"Please, try using ISO3 version with BDIFTypes.FLAG_NON_STRICT_READ flag to remove non-unique minutiae while reading.");
			}

			// Convert FMCRecord to NFRecord
			NFRecord nfRecord = fmcRecord.toNFRecord();

			// Add NFRecord to NFTemplate
			NFTemplate nfTemplate = new NFTemplate();
			nfTemplate.getRecords().add(nfRecord);
			// Set NFTemplate to NTemplate
			nTemplate = new NTemplate();
			nTemplate.setFingers(nfTemplate);

			// Pack NTemplate object
			packedNTemplate = nTemplate.save();
			NFile.writeAllBytes(args[1], packedNTemplate);
		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			if (fmcRecord != null) fmcRecord.dispose(); // explicit resource clean-up
			if (nTemplate != null) nTemplate.dispose();
			if (storedFmcRecord != null) storedFmcRecord.dispose();
			if (packedNTemplate != null) packedNTemplate.dispose();
		}
	}
}

File: Tutorials/BiometricStandards/C/FMCRecordToNTemplate/FMCRecordToNTemplate.c
#include <TutorialUtils.h>

#ifdef N_APPLE_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NBiometrics/NBiometrics.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NBiometrics.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("FMCRecordToNTemplate");
const NChar description[] = N_T("Demonstrates creation of NTemplate from FMCRecord");
const NChar version[] = N_T("13.1.0.0");
const NChar copyright[] = N_T("Copyright (C) 2017-2024 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [FMCRecord] [NTemplate] [Version] [MinutiaeFormat] [ReadBerTlvDo]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tFMCRecord - filename with FMCRecord\n"));
	printf(N_T("\tNTemplate - filename for NTemplate\n"));
	printf(N_T("\tVersion - FMCRecord version\n"));
	printf(N_T("\t\tISO2 - ISO/IEC 19794-2:2005\n"));
	printf(N_T("\t\tISO3 - ISO/IEC 19794-2:2011\n"));
	printf(N_T("\tMinutiaeFormat - card format of minutiae data\n"));
	printf(N_T("\t\tC - compact size\n"));
	printf(N_T("\t\tN - normal size (ISO2 only)\n"));
	printf(N_T("\tReadBerTlvDo - FMCRecord read from buffer option\n"));
	printf(N_T("\t\t1 - Biometric Data Template (BDT) or Cardholder\n"));
	printf(N_T("\t\t    Biometric Data (CBD) BER-TLV DO (Data Object)\n"));
	printf(N_T("\t\t0 - minutiae data buffer only.\n"));
	printf(N_T("example:\n"));
	printf(N_T("\t%s fmcRecord.FMCRecord template.NTemplate ISO3 C 1\n"), title);

	return 1;
}

NResult StoredFMCRecordToPackedNFRecord(
	HNBuffer hBuffer, // buffer that contains FMCRecord
	HNBuffer * phBuffer,// buffer that contains packed NTemplate
	BdifStandard standard,
	NVersion standardVersion,
	FmcrMinutiaFormat minutiaFormat,
	NBool isBerTlv,
	NUInt flags
)
{
	HFMCRecord hFMCRecord = NULL; // handle to FMCRecord object
	HNFRecord hNFRecord = NULL; // handle to NFRecord object
	HNTemplate hNTemplate = NULL; // handle to NTemplate object
	HNFTemplate hNFTemplate = NULL; // handle to NFTemplate object
	NResult result;

	if (isBerTlv)
	{

		// Create FMCRecord object from FMCRecord BER-TLV Data Object (DO) stored in memory (e.g., BDT or CBD BER-TLV DO)
		result = FMCRecordCreateFromMemoryN(hBuffer, standard, standardVersion, minutiaFormat, flags, NULL, &hFMCRecord);
		if(NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("error in FMCRecordCreateFromMemory, error code: %d\n"), result);
			goto FINALLY;
		}
	}
	else if (!isBerTlv)
	{
		// Create FMCRecord object
		result = FMCRecordCreate(standard, standardVersion, minutiaFormat, 0, &hFMCRecord);
		if(NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("error in FMCRecordCreate, error code: %d\n"), result);
			goto FINALLY;
		}

		// Read minutiae data from buffer to FMCRecord
		result = FMCRecordSetMinutiaeBuffer(hFMCRecord,  hBuffer);
		if(NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("error in FMCRecordSetMinutiaeBuffer, error code: %d\n"), result);
			goto FINALLY;
		}
	}
	// Explicitly check minutiae uniqueness for FMCRecord v2.0 as well (since mandatory implicit check since v3.0 only)
	if (standardVersion == FMCR_VERSION_ISO_2_0)
	{
		NBool isUnique = NFalse;
		result = FMCRecordValidateMinutiaeUniqueness(hFMCRecord, &isUnique);
		if(NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("error in FMCRecordValidateMinutiaeUniqueness, error code: %d\n"), result);
			goto FINALLY;
		}
		if (!isUnique)
		{
			printf(N_T("!!!WARNING!!!:\nNot all minutiae in FMCRecord are unique!\nPlease, try using ISO3 version with BDIF_NON_STRICT_READ flag to remove non-unique minutiae while reading.\n"));
		}
	}

	// Convert FMCRecord to NFRecord
	result = FMCRecordToNFRecord(hFMCRecord, 0, &hNFRecord);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in FMCRecordToNFRecord, error code: %d\n"), result);
		goto FINALLY;
	}

	// Create NTemplate
	result = NTemplateCreateEx(0, &hNTemplate);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NTemplateCreate() failed (result = %d)!"), result);
		goto FINALLY;
	}

	// Create NFTemplate
	result = NFTemplateCreateEx(NFalse, 0, &hNFTemplate);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NFTemplateCreateEx() failed (result = %d)!"), result);
		goto FINALLY;
	}

	// Set NFTemplate to NTemplate
	result = NTemplateSetFingers(hNTemplate, hNFTemplate);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NTemplateSetFingers() failed (result = %d)!"), result);
		goto FINALLY;
	}

	// Add NFRecord to NFTemplate
	result = NFTemplateAddRecordEx(hNFTemplate, hNFRecord, NULL);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NFTemplateAddRecordEx() failed (result = %d)!"), result);
		goto FINALLY;
	}

	// Pack NTemplate object
	result = NObjectSaveToMemoryN(hNTemplate, 0, phBuffer);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NObjectSaveToMemoryN, error code: %d\n"), result);
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hFMCRecord);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hNFRecord);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hNFTemplate);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hNTemplate);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
	}

	return result;
}

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	HNBuffer hBuffer = NULL;
	HNBuffer hBufferIn = NULL;
	NBool available = NFalse;
	BdifStandard standard = bsIso;
	NVersion standardVersion;
	FmcrMinutiaFormat minutiaFormat;
	NUInt flags = 0; // BDIF_NON_STRICT_READ - removes non-unique minutiae for FMCR_VERSION_ISO_3_0 only
	NBool isBerTlv;
	
	//=========================================================================
	// CHOOSE LICENCES !!!
	//=========================================================================
	// ONE of the below listed licenses is required for unlocking this sample's functionality. Choose a license that you currently have on your device.
	// If you are using a TRIAL version - choose any of them.

	const NChar * license = N_T("FingerClient");
	//const NChar * license = N_T("FingerFastExtractor");

	//=========================================================================

	OnStart(title, description, version, copyright, argc, argv);

	if(argc < 6)
	{
		OnExit();
		return usage();
	}

	//=========================================================================
	// TRIAL MODE
	//=========================================================================
	// Below code line determines whether TRIAL is enabled or not. To use purchased licenses, don't use below code line.
	// GetTrialModeFlag() method takes value from "Bin/Licenses/TrialFlag.txt" file. So to easily change mode for all our examples, modify that file.
	// Also you can just set TRUE to "TrialMode" property in code.
	{
		NBool isTrialMode = NFalse;
		result = GetTrialModeFlag(&isTrialMode);
		if (NFailed(result))
		{
			goto FINALLY;
		}
		result = NLicManSetTrialMode(isTrialMode);
		if (NFailed(result))
		{
			goto FINALLY;
		}
		printf(N_T("Trial mode: %d\n"), (int)isTrialMode);
	}
	//=========================================================================

	// check the license first
	result = NLicenseObtain(N_T("/local"), N_T("5000"), license, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtain() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("License %s not available\n"), license);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	if (!strcmp(argv[3], N_T("ISO2")))
	{
		standardVersion = FMCR_VERSION_ISO_2_0;
	}
	else if (!strcmp(argv[3], N_T("ISO3")))
	{
		standardVersion = FMCR_VERSION_ISO_3_0;
	}
	else
	{
		printf(N_T("wrong version!"));
		result = N_E_FAILED;
		goto FINALLY;
	}

	if (!strcmp(argv[4], N_T("C")))
	{
		minutiaFormat = fmcrmfCompactSize;
	}
	else if (!strcmp(argv[4], N_T("N")))
	{
		minutiaFormat = fmcrmfNormalSize;
	}
	else
	{
		printf(N_T("wrong minutia format!"));
		result = N_E_FAILED;
		goto FINALLY;
	}

	// check if full FMCRecord BER-TLV Data Object (DO) (e.g., Biometric Data Template (BDT) or Cardholder Biometric Data (CBD)) or minutiae buffer only is available
	isBerTlv = !strcmp(argv[5], N_T("1")) ? NTrue : NFalse;

	result = NFileReadAllBytesCN(argv[1], &hBufferIn);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NFileReadAllBytesCN, error code: %d\n"), result);
		goto FINALLY;
	}

	result = StoredFMCRecordToPackedNFRecord(hBufferIn, &hBuffer, standard, standardVersion, minutiaFormat, isBerTlv, flags);
	if(NFailed(result))
	{
		goto FINALLY;
	}

	result = NFileWriteAllBytesCN(argv[2], hBuffer);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NFileWriteAllBytesCN, error code: %d\n"), result);
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hBuffer);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hBufferIn);
		if (NFailed(result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
	}

	OnExit();
	return result;
}
